File: 2-7_mq-exporter/mq_exporter.go
// ./mq_exporter --port 8000 --interval 15
// curl http://localhost:8000/metrics

package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

type QueueMetrics struct {
	Name               string  `json:"name"`
	Vhost              string  `json:"vhost"`
	Messages           float64 `json:"messages"`
	MessagesReady      float64 `json:"messages_ready"`
	MessagesUnacknowledged float64 `json:"messages_unacknowledged"`
}

type Exporter struct {
	rabbitMQHost     string
	rabbitMQUser     string
	rabbitMQPassword string
	client           *http.Client

	messages           *prometheus.Desc
	messagesReady      *prometheus.Desc
	messagesUnacknowledged *prometheus.Desc
}

func NewExporter(host, user, password string) *Exporter {
	return &Exporter{
		rabbitMQHost:     host,
		rabbitMQUser:     user,
		rabbitMQPassword: password,
		client:           &http.Client{Timeout: 10 * time.Second},

		messages: prometheus.NewDesc(
			"rabbitmq_individual_queue_messages",
			"Total number of messages in RabbitMQ queue",
			[]string{"host", "vhost", "name"},
			nil,
		),
		messagesReady: prometheus.NewDesc(
			"rabbitmq_individual_queue_messages_ready",
			"Total number of ready messages in RabbitMQ queue",
			[]string{"host", "vhost", "name"},
			nil,
		),
		messagesUnacknowledged: prometheus.NewDesc(
			"rabbitmq_individual_queue_messages_unacknowledged",
			"Total number of unacknowledged messages in RabbitMQ queue",
			[]string{"host", "vhost", "name"},
			nil,
		),
	}
}

func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
	ch <- e.messages
	ch <- e.messagesReady
	ch <- e.messagesUnacknowledged
}

func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
	url := fmt.Sprintf("http://%s/api/queues", e.rabbitMQHost)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		log.Printf("Failed to create HTTP request: %v", err)
		return
	}
	req.SetBasicAuth(e.rabbitMQUser, e.rabbitMQPassword)

	resp, err := e.client.Do(req)
	if err != nil {
		log.Printf("Failed to fetch data from RabbitMQ API: %v", err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		log.Printf("Unexpected status code from RabbitMQ API: %d", resp.StatusCode)
		return
	}

	var queues []QueueMetrics
	if err := json.NewDecoder(resp.Body).Decode(&queues); err != nil {
		log.Printf("Failed to decode RabbitMQ API response: %v", err)
		return
	}

	host := e.rabbitMQHost
	for _, queue := range queues {
		ch <- prometheus.MustNewConstMetric(
			e.messages,
			prometheus.GaugeValue,
			queue.Messages,
			host, queue.Vhost, queue.Name,
		)
		ch <- prometheus.MustNewConstMetric(
			e.messagesReady,
			prometheus.GaugeValue,
			queue.MessagesReady,
			host, queue.Vhost, queue.Name,
		)
		ch <- prometheus.MustNewConstMetric(
			e.messagesUnacknowledged,
			prometheus.GaugeValue,
			queue.MessagesUnacknowledged,
			host, queue.Vhost, queue.Name,
		)
	}
}

func main() {
	rabbitMQHost := os.Getenv("RABBITMQ_HOST")
	rabbitMQUser := os.Getenv("RABBITMQ_USER")
	rabbitMQPassword := os.Getenv("RABBITMQ_PASSWORD")

	if rabbitMQHost == "" || rabbitMQUser == "" || rabbitMQPassword == "" {
		log.Fatal("Environment variables RABBITMQ_HOST, RABBITMQ_USER, and RABBITMQ_PASSWORD must be set")
	}

	port := flag.String("port", "8000", "Port to expose metrics")
	interval := flag.Int("interval", 15, "Interval to scrape RabbitMQ metrics (in seconds)")
	flag.Parse()

	exporter := NewExporter(rabbitMQHost, rabbitMQUser, rabbitMQPassword)
	prometheus.MustRegister(exporter)

	http.Handle("/metrics", promhttp.Handler())
	serverAddr := fmt.Sprintf(":%s", *port)

	log.Printf("Starting RabbitMQ exporter on %s (scraping every %d seconds)", serverAddr, *interval)
	log.Fatal(http.ListenAndServe(serverAddr, nil))
}

File: mq_exporter.py
from prometheus_client import start_http_server, Gauge
import requests
import time
import argparse

total_messages_gauge = Gauge(
    'rabbitmq_queue_messages_total',
    'Total number of messages in queues.'
    ['vhost', 'queue']
)

messages_ready_gauge = Gauge(
    'rabbitmq_queue_messages_ready',
    'Number of ready messages in queues.'
    ['vhost', 'queue']
)

messages_unacknowledged_gauge = Gauge(
    'rabbitmq_queue_messages_total',
    'Number of unacknowledged messages in queues.'
    ['vhost', 'queue']
)

def fetch_rabbitmq_metrics(api_url, username, password):
    try:
        response = requests.get(f'{api_url}/queues', auth=(username, password))
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"Error fetching RabbitMQ metrics: {e}")
        return []
    
def update_metrics(data):
    for queue in data:
        vhost = queue['vhost']
        name = queue['name']
        total_messages = queue.get('messages', 0)
        messages_ready = queue.get('message_ready', 0)
        messages_unacknowledged = queue.get('messages_unacknowledged', 0)

        total_messages_gauge.labels(vhost=vhost, queue=name).set(total_messages)
        messages_ready_gauge.labels(vhost=vhost, queue=name).set(messages_ready)
        messages_unacknowledged_gauge.labels(vhost=vhost, queue=name).set(messages_unacknowledged)

def main():
    parser = argparse.ArgumentParser(description='RabbitMQ Prometheus Exporter')
    parser.add_argument('--api-url', required=True, help='RabbitMQ HTTP APU URL')
    parser.add_argument('--username', required=True, help='RabbitMQ username')
    parser.add_argument('--password', required=True, help='RabbitMQ password')
    parser.add_argument('--port', type=int, default=8000, help='Exposed port to Prometheus')
    parser.add_argument('--interval', type=int, default=15, help='Interval to seconds to fetch metrics')

    args = parser.parse_args()

    start_http_server(args.port)
    print(f"Starting Prometheus exporter on port {args.port}...")

    while True:
        data = fetch_rabbitmq_metrics(args.api_url, args.username, args.password)
        update_metrics(data)
        time.sleep(args.interval)

if __name__ == '__main__':
    main()